examples/cpp/src/cagra_persistent_example.cu

```diff
@@ -0,0 +1,256 @@
+        auto work_size = queries.extent(0);
+        using index_type = typeof(work_size);
+        // Limit the maximum number of concurrent jobs
+        constexpr index_type kMaxJobs = 1000;
+        std::array<std::future<void>, kMaxJobs> futures;
+        for (index_type i = 0; i < work_size + kMaxJobs; i++) {
+          // wait for previous job in the same slot to finish
+          if (i >= kMaxJobs) {
+            futures[i % kMaxJobs].wait();
+          }
+          // submit a new job
+          if (i < work_size) {
+            futures[i % kMaxJobs] = std::async(std::launch::async, [&]() {
+              cagra::search(res, ps, index, slice_matrix(queries, i, 1),
+                            slice_matrix(neighbors, i, 1),
+                            slice_matrix(distances, i, 1));
+            });
+          }
+        }
+        /* See the remark for search_batch */
+        if (needs_sync) {
+          raft::resource::sync_stream(res);
+        }
+      };
+
+  // Launch the baseline search: check the big-batch performance
+  time_it("standard/batch A", search_batch, true, search_params, queries_a,
+          neighbors_a, distances_a);
+  time_it("standard/batch B", search_batch, true, search_params, queries_b,
+          neighbors_b, distances_b);
+
+  // Try to handle the same amount of work in the async setting using the
+  // standard implementation.
+  // (Warning: suboptimal - it uses a single stream for all async jobs)
+  time_it("standard/async A", search_async, true, search_params, queries_a,
+          neighbors_a, distances_a);
+  time_it("standard/async B", search_async, true, search_params, queries_b,
+          neighbors_b, distances_b);
+
+  // Do the same using persistent kernel.
+  time_it("persistent/async A", search_async, false, search_params_persistent,
+          queries_a, neighbors_a, distances_a);
+  time_it("persistent/async B", search_async, false, search_params_persistent,
+          queries_b, neighbors_b, distances_b);
+  /*
+Here's an example output:
+```
+CAGRA index has 1000000 vectors
+CAGRA graph has degree 64, graph size [1000000, 64]
+[standard/batch A] execution time: 854.645 ms
+[standard/batch B] execution time: 698.58 ms
+[standard/async A] execution time: 19190.6 ms
+[standard/async B] execution time: 18292 ms
+[I] [15:56:49.756754] Initialized the kernel 0x7ea4e55a5350 in stream
+              139227270582864; job_queue size = 8192; worker_queue size = 155
+[persistent/async A] execution time: 1285.65 ms
+[persistent/async B] execution time: 1316.97 ms
+[I] [15:56:55.756952] Destroyed the persistent runner.
+```
+Note, the persistent kernel time in async mode (1 query per job) is up to 2x
+slower than the standard kernel with the huge batch size (100K queries).
+One reason for this is the non-optimal CTA size: CAGRA kernels are automatically
+tuned for latency and so use large CTA sizes when the batch size is small.
+Try explicitly setting the search parameter `thread_block_size` to a small
```

Comment:
It is, and that is a problem. When `n_queries == 1`, the block size is big to improve latency; but that hurts the throughput of the persistent kernel when a big number of queries is actually processes in parallel.

Reply:
So if it is calculated using `persistent_device_usage`, then does `thread_block_size` have an effect?

Reply:
Ah, I misunderstood the last bit of your question. These are orthogonal parameters. `thread_block_size` is the CTA size and it affects how many CTAs can be launched within a single SM. It's calculated from the problem size automatically, unless specified explicitly (in the same way for both standard and persistent kernel). This determines the maximum resident grid size.
`persistent_device_usage` lets user reduce the grid size, so that one or few SMs stay idle during kernel execution.